quinn/src/streams.rs

```diff
@@ -479,46 +539,88 @@ where
         self.stream
     }
 
+    /// Handle common logic related to reading out of a receive stream
+    ///
+    /// This takes an `FnMut` closure that takes care of the actual reading process, matching
+    /// the detailed read semantics for the calling function with a particular return type.
+    /// The closure can read from the passed `&mut Chunks` and has to return a result in two parts:
+    ///
+    /// * Amount of data read, in the type expected by the calling read method
+    /// * Status after the final read, with the `Ok` value denoting whether the stream has finished
     fn poll_read_generic<T, U>(
         &mut self,
         cx: &mut Context,
+        ordered: bool,
         mut read_fn: T,
     ) -> Poll<Result<Option<U>, ReadError>>
     where
-        T: FnMut(
-            &mut crate::connection::ConnectionInner<S>,
-            StreamId,
-        ) -> Result<Option<U>, proto::ReadError>,
+        T: FnMut(&mut Chunks) -> (Option<U>, Result<ReadStatus, proto::ReadError>),
```

Comment:
Is `(None, Ok(Readable))` a valid state? Should this be `(Option<U>, Result<(), proto::ReadError>)`?

Reply:
Valid states:
- (Some, Ok(Readable))
- (Some, Ok(Finished))
- (Some, Err)
- (None, Ok(Finished))
- (None, Err)

No, that wouldn't work because `(Some, Ok(Finished))` is valid. How about this?
```
enum ReadResult<U> {
  Nonempty(U, Result<ReadStatus, proto::ReadError>),
  Empty(Result<(), proto::ReadError>),
}
```

(terrible naming, please improve)

Reply:
Better yet, I think this is algebraically equivalent, but much saner:
```rust
enum ReadResult<U> {
    Readable(U),
    Finished(Option<U>),
    Failed(Option<U>, proto::ReadError),
}
```

Reply:
The latter does not seem like an improvement:

```diff
diff --git a/quinn/src/streams.rs b/quinn/src/streams.rs
index 9e7fcfef..317b973a 100644
--- a/quinn/src/streams.rs
+++ b/quinn/src/streams.rs
@@ -381,13 +381,7 @@ where
             let mut read = false;
             loop {
                 if buf.remaining() == 0 {
-                    return (
-                        match read {
-                            true => Some(()),
-                            false => None,
-                        },
-                        Ok(ReadStatus::Readable),
-                    );
+                    return ReadStatus::Readable(());
                 }
 
                 match chunks.next(buf.remaining()) {
@@ -395,13 +389,19 @@ where
                         buf.put_slice(&chunk.bytes);
                         read = true;
                     }
-                    res => {
-                        return (
+                    Ok(None) => {
+                        return ReadStatus::Finished(match read {
+                            true => Some(()),
+                            false => None,
+                        })
+                    }
+                    Err(e) => {
+                        return ReadStatus::Failed(
                             match read {
                                 true => Some(()),
                                 false => None,
                             },
-                            res.map(|inner| inner.into()),
+                            e,
                         )
                     }
                 }
@@ -554,7 +554,7 @@ where
         mut read_fn: T,
     ) -> Poll<Result<Option<U>, ReadError>>
     where
-        T: FnMut(&mut Chunks) -> (Option<U>, Result<ReadStatus, proto::ReadError>),
+        T: FnMut(&mut Chunks) -> ReadStatus<U>,
     {
         use proto::ReadError::*;
         if self.all_data_read {
@@ -568,26 +568,25 @@ where
 
         // If we stored an error during a previous call, return it now. This can happen if a
         // `read_fn` both wants to return data and also returns an error in its final stream status.
-        let (read, result) = match self.reset.take() {
-            Some(code) => (None, Err(Reset(code))),
+        let status = match self.reset.take() {
+            Some(code) => ReadStatus::Failed(None, Reset(code)),
             None => {
                 let mut chunks = conn.inner.read(self.stream, ordered)?;
-                let (read, result) = read_fn(&mut chunks);
+                let status = read_fn(&mut chunks);
                 if chunks.finalize().should_transmit() {
                     conn.wake();
                 }
-                (read, result)
+                status
             }
         };
 
-        match result {
-            Ok(status) => {
-                if let ReadStatus::Finished = status {
-                    self.all_data_read = true;
-                }
+        match status {
+            ReadStatus::Readable(read) => Poll::Ready(Ok(Some(read))),
+            ReadStatus::Finished(read) => {
+                self.all_data_read = true;
                 Poll::Ready(Ok(read))
             }
-            Err(Blocked) => match read {
+            ReadStatus::Failed(read, Blocked) => match read {
                 Some(val) => Poll::Ready(Ok(Some(val))),
                 None => {
                     if let Some(ref x) = conn.error {
@@ -597,7 +596,7 @@ where
                     Poll::Pending
                 }
             },
-            Err(Reset(error_code)) => match read {
+            ReadStatus::Failed(read, Reset(error_code)) => match read {
                 None => {
                     self.all_data_read = true;
                     Poll::Ready(Err(ReadError::Reset(error_code)))
@@ -611,18 +610,10 @@ where
     }
 }
 
-enum ReadStatus {
-    Readable,
-    Finished,
-}
-
-impl<T> From<Option<T>> for ReadStatus {
-    fn from(val: Option<T>) -> Self {
-        match val {
-            Some(_) => ReadStatus::Readable,
-            None => ReadStatus::Finished,
-        }
-    }
+enum ReadStatus<T> {
+    Readable(T),
+    Finished(Option<T>),
+    Failed(Option<T>, proto::ReadError),
 }
 
 /// Future produced by [`RecvStream::read_to_end()`].
```

Reply:
Okay, I figured out how to make it work. Should be good.